Update generic_thermostat.py

MrMep opened the pull request:
After _control_heating() is executed, current_operation() is correctly called but _is_device_active() still reports the old state of the heater switch, at least in my case. The resulting climate state is incorrect until the next refresh, which apparently occurs only when _sensor_changed() gets called (it can be minutes after).

I think the state of the heater switch should be forced to update at the end of  _control_heating(), but I don't know how to do that...
A simple sleep() fixes it, but obviously it's just a temporary workaround, I'm not really expecting this PR to be actually merged, unless there's no other solution.

**Description:**


**Related issue (if applicable):** fixes #5418

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#<home-assistant.github.io PR number goes here>

**Example entry for `configuration.yaml` (if applicable):**
```yaml

```

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51

MrMep committed:
Update generic_thermostat.py

After _control_heating() is executed, current_operation() is correctly called but _is_device_active() still reports the old state if the heater switch, at least in my case. The resulting climate state is incorrect until the next refresh, which apparently occurs only when _sensor_changed() gets called (it can be minutes after).

I think the state of the heater switch should be forced to update at the end of  _control_heating(), but I don't know how to do that...
A simple sleep() fixes it, but obviously is just a temporary workaround, I'm not really expecting this PR to be actually committed, unless there's no other solution.

MrMep committed:
Update generic_thermostat.py

armills commented:
First off, just as an FYI of what's happening in the background, after `_control_heating` is called, [`self.update_ha_state`](https://github.com/MrMep/home-assistant/blob/72dca1da091945651788ed6cc5e314570b79361f/homeassistant/components/climate/generic_thermostat.py#L138) is called. This method reads all of the properties for this component. (This is where current_operation is read from)

[Try adding a new callback like this line with a new function `self._switch_changed`](https://github.com/MrMep/home-assistant/blob/72dca1da091945651788ed6cc5e314570b79361f/homeassistant/components/climate/generic_thermostat.py#L90)

[You can duplicate this function, and leave only the line`self.schedule_update_ha_state()`](https://github.com/MrMep/home-assistant/blob/72dca1da091945651788ed6cc5e314570b79361f/homeassistant/components/climate/generic_thermostat.py#L160)

MrMep commented:
Thank you!
I hadn't thought of that!
I did that, but it doesn't kill it yet... 
I'll update the file and post my comments

MrMep committed:
Update generic_thermostat.py

As suggested, I added a callback on the heater switch state change.
Still it doesn't solve the problem.

MrMep commented:
So, as @armills suggested, I added a `track_state_change()` on the heater switch and the corresponding call back `_switch_changed()`
It gets a little better, but still it doesn't solve the problem: if you change the target temperature (respecting `min_cycle_duration`), the thermostat works well, but the state is incorrect.
I'm not sure, but it's like `_switch_changed()` gets called correctly, but then `switch.is_on()` still reports the old state.
Is it possible that `track_state_change()` triggers in a timely manner, while `switch.is_on()` doesn't?
I don't know well enough HA internal workings to verify that.

armills commented:
Hmm, good to know that we're at least getting better. Does it report incorrectly in both directions? i.e.

1) Change the target, switch turns off, state still shows active.
2) Change the target, switch turns on, state still shows inactive.

armills commented:
Can you add some debug logs to [`current_operation`](https://github.com/MrMep/home-assistant/blob/a87d653077b0a83c6fef29854694d1f225e50804/homeassistant/components/climate/generic_thermostat.py#L117) and print out all the variables used to make the determination?

MrMep commented:
I did that. The fact is that `update_ha_state` gets called first from `set_temperature`: there `current_operation` has the wrong values. Then it gets called by '_switch_changed', and there it has the right values, but it's like the first call (the wrong one) gets executed last.

The report is incorrect both ways, but the error is sort of random: sometimes it works, sometimes not. I think it depends on execution time: when I put some heavier debug lines, everything was working perfectly, I guess because of the slight slowing down.

Maybe it has to do with some async voodoo the guys did in entity_id helpers?

In summary, the component now seems to be working correctly, but the state is randomly wrongly reported when the target temperature is changed.

armills commented:
I see. Changing the `self.update_ha_state()` in `set_temperature` to `self.schedule_update_ha_state()` should work. Because `update_ha_state` will block the calling thread until complete, it is given higher priority in the event loop, so the scheduled update happens later. Since `set_temperature` doesn't need to block until the update is complete, we can just change it.

MrMep commented:
Thanks!
Actually I had tried that too. I retried and even moved `self.schedule_update_ha_state()` before `self._control_heating()` in `set_temperature(),` but unfortunately nothing changes...

When the error occurs, I can actually see the state quickly change twice in the frontend: the first one is correct, then the wrong one.

armills commented:
So you said `current_operation` has the wrong values. Which ones are wrong? I think you'll have to drill into those. It really shouldn't matter which order they are processed, since the state will have already changed by the second execution.

MrMep commented:
It's always `self._is_device_active` to be wrong, even because it looks to me that `self._active`, once initialized, remains always active (but that's a topic for another discussion, maybe)

armills commented:
[OK, the `is_device_active` is coming from the switch state itself.](https://github.com/MrMep/home-assistant/blob/a87d653077b0a83c6fef29854694d1f225e50804/homeassistant/components/climate/generic_thermostat.py#L234)

In that case, can you include a logger output on `current_operation`, and paste your logs from the first `current_operation` call to the second? There must be a state change somewhere in-between. What switch platform are you using?

MrMep commented:
Here it is:

```
19:16:59 INFO (Thread-3) [homeassistant.components.climate.generic_thermostat] DEBUGGING: Termostato Bagno - set_temperature 36.0 
19:16:59 INFO (MainThread) [homeassistant.components.climate.generic_thermostat] DEBUGGING: Termostato Bagno -current_operation - State is False - Clima is: True - Device is: False
19:16:59 INFO (MainThread) [homeassistant.components.climate] DEBUGGING: state - Calling current_operation Termostato Bagno
19:16:59 INFO (MainThread) [homeassistant.components.climate.generic_thermostat] DEBUGGING: Termostato Bagno -current_operation - State is False - Clima is: True - Device is: False
19:16:59 INFO (MainThread) [homeassistant.components.climate.generic_thermostat] DEBUGGING: Termostato Bagno -current_operation - State is False - Clima is: True - Device is: False
19:16:59 INFO (Thread-3) [homeassistant.components.climate.generic_thermostat] DEBUGGING: Termostato Bagno -_control_heating
19:16:59 INFO (Thread-7) [homeassistant.components.climate.generic_thermostat] DEBUGGING: Termostato Bagno -_switch_changed
19:16:59 INFO (MainThread) [homeassistant.components.climate.generic_thermostat] DEBUGGING: Termostato Bagno -current_operation - State is True - Clima is: True - Device is: True
19:16:59 INFO (MainThread) [homeassistant.components.climate] DEBUGGING: state - Calling current_operation Termostato Bagno
19:16:59 INFO (MainThread) [homeassistant.components.climate.generic_thermostat] DEBUGGING: Termostato Bagno -current_operation - State is True - Clima is: True - Device is: True
19:16:59 INFO (MainThread) [homeassistant.components.climate.generic_thermostat] DEBUGGING: Termostato Bagno -current_operation - State is True - Clima is: True - Device is: True
```

![schermata 2017-01-19 alle 19 17 03](https://cloud.githubusercontent.com/assets/11885198/22119529/36358daa-de7c-11e6-8405-3be0f1759e2f.jpg)

MrMep commented:
Mines are MQTT switches.
As you can see, I think the sequence is logical and correct: current_operation gets called (trice! I don't know why) first from set_temperature: the switch state isn't updated yet, so it gives an incorrect response.
But then _switch_changed calls another update, and current_operation this time gives the right response. 
But in the frontend the state is wrong.

armills commented:
I see, so after this happens, try refreshing your browser, does it show the correct state then?

MrMep commented:
Yeah, whatever happens after that, the state is correct, unless it comes from set_temperature and the switch gets toggled.

The point is that the frontend gets updated perfectly and timely always, I can see the double update with my very eyes, even if it's quick.

Sometimes, when the target temperature is changed and thus the switch is toggled, there's a false response.

Ok, it's not important now that we added _switch_changed(), but still... I would like to know why...

armills commented:
In that case, I think the code here was necessary to get the correct status into HA, but there's a frontend bug that's causing the statuses to get delivered out-of-order, do you agree?

MrMep committed:
Update generic_thermostat.py

MrMep commented:
It looks like it...

I made a few minor changes, I swapped   self.schedule_update_ha_state() and  self._control_heating() in _sensor_changed and set_temperature: it makes more sense now that we have _switch_changed.

As long as I'm concerned, this can be merged and closed.

One more thing though, bear with me: having self.is_active always true after initialization basically means that we can't control the heater  independently: even if we toggle the device switch directly, the thermostat is bound to take control at the first temp sensor change, do you agree?

Are we sure of this? Many hw thermostats (but not all of them) give the option to manually switch on/off the heater, regardless of the target temperature.

What is the best place to discuss this, do you think?

Really thanks for all, I've learned a lot :)

armills commented:
self.schedule_update_ha_state() should still be after self._control_heating(). We don't want to update the state in HA until we're done taking actions, but otherwise this looks good to me. I agree that being able to temporarily disable the thermostat is a good idea for a new feature, but that's out of scope for this fix. It would be similar to how automations can be toggled now.

A good place to start the discussion is the developer chat: https://gitter.im/home-assistant/home-assistant/devs For something more permanent you can start a thread in the Feature Request forum. Beyond that the most effective way to start the conversation is just open a PR with the new functionality. :)

MrMep commented:
Thanks for the hints :)

As for the order of functions, that's my reasoning: in set_temperature and _sensor_changed we need to update the state with the new attribute value. Then,  the result of _control_heating is either the switch is toggled or not: in the first case _switch_changed gets called and it updates the state again, in the second case the state doesn't change.
Maybe it's a question of taste: it seemed more logical to me, or am I missing something?
For sure, if we didn't need _switch_changed in the first place, it would be better and I would agree with you 100%

armills commented:
Mostly it's just a matter of style so that the next person to look at this doesn't need to figure out why we're updating state in the middle of execution. Most all updates happen at the end of every function by convention. We also need to ensure that update_ha_state is called even when the switch state doesn't change, because the temperature target needs to be updated.

Also, just as an FYI, when we call `switch.turn_on`, we're calling from an executor thread, and the actual event to turn the switch on is queued for execution in the main thread, so it doesn't happen right away.

MrMep commented:
Clear, thank you, I'm reverting the swap.

MrMep committed:
Update generic_thermostat.py

emlove reviewed (approved):
LGTM, I just want another dev to take a look with fresh eyes.

Danielhiversen commented:
LGTM
